core/src/main/scala/org/apache/spark/shuffle/FetchFailedException.scala

```diff
@@ -45,6 +45,12 @@ private[spark] class FetchFailedException(
     this(bmAddress, shuffleId, mapId, reduceId, cause.getMessage, cause)
   }
 
+  // SPARK-19267. We set the fetch failure in the task context, so that even if there is user-code
+  // which intercepts this exception (possibly wrapping it), the Executor can still tell there was
+  // a fetch failure, and send the correct error msg back to the driver.  The TaskContext won't be
+  // defined if this is run on the driver (just in test cases) -- we can safely ignore then.
+  Option(TaskContext.get()).map(_.setFetchFailed(this))
```

Comment:
yes, good point.  I added to the docs, does it look OK?

I also considered making the call to `TaskContext.setFetchFailed` live outside of the constructor, so at each site it was created, it would have to be called -- but I thought that seemed more dangerous.